### PR TITLE
upgrade connectedk8s to use Microsoft Graph

### PR DESCRIPTION
**Pull request by TheOnlyWei**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### Related command
az connectedk8s

### General Guidelines

- [X] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [X] Have you run `python scripts/ci/test_index.py -q` locally? (`pip install wheel==0.30.0` required)

### CODE REVIEW

**Comment by yonzhan** [comment]
Thank you for your contribution! We will review the pull request and get back to you soon.

File: src/connectedk8s/HISTORY.rst
Comment (NarayanThiru) [inline src/connectedk8s/HISTORY.rst:5]: @TheOnlyWei , can you share why you are making this change?  Can you update the description with this explanation and link to the ADO Task tracking this work?

File: src/connectedk8s/azext_connectedk8s/custom.py
Comment (NarayanThiru) [inline src/connectedk8s/azext_connectedk8s/custom.py:32]: With this change to use MS Graph, does the customer/client need different permissions?  Specifically, is it possible for this to fail, for someone for whom it used to work before?  This will become a regression - hence asking.

**Review by NarayanThiru — CHANGES_REQUESTED** [review]

File: src/connectedk8s/HISTORY.rst
Comment (TheOnlyWei) [reply src/connectedk8s/HISTORY.rst:5]: Replied in Teams chat with details.

File: src/connectedk8s/azext_connectedk8s/custom.py
Comment (TheOnlyWei) [reply src/connectedk8s/azext_connectedk8s/custom.py:32]: The customer will need MS Graph permission for the extension to get the object ID for them if they don't want to pass it in. There is no way around this in an automated way without using MS Graph, since AAD Graph is deprecated. The behavior will be the same, in that if they don't have permission, it will try to use the one passed in, and if the customer didn't pass anything in, then Custom Locations RP won't be enabled. The AAD Graph delegated user permissions should be automatically converted to the equivalent MS Graph user permissions. This call specifically uses the list API, and the permissions are specified below:

https://learn.microsoft.com/en-us/graph/api/serviceprincipal-list?view=graph-rest-1.0&tabs=http#permissions

File: src/connectedk8s/HISTORY.rst
Comment (bganapa) [inline src/connectedk8s/HISTORY.rst:6]: Version needs update

File: src/connectedk8s/setup.py
Comment (bganapa) [inline src/connectedk8s/setup.py:20]: version update

File: src/connectedk8s/azext_connectedk8s/custom.py
Comment (bganapa) [inline src/connectedk8s/azext_connectedk8s/custom.py:2201]: CLI is already making use of MS graph in other places, you are just reusing the same here . Is that correct? @TheOnlyWei

File: src/connectedk8s/HISTORY.rst
Comment (bganapa) [reply src/connectedk8s/HISTORY.rst:5]: Lets use the PR for discussion @TheOnlyWei for why the change is needed

**Comment by bganapa** [comment]
The PR needs @julienstroheker's approval. Please dont merge it without his approval

**Comment by bganapa** [comment]
@TheOnlyWei Can you please list what manual validations are done against azure to cover the changes?

File: src/connectedk8s/HISTORY.rst
Comment (TheOnlyWei) [reply src/connectedk8s/HISTORY.rst:19]: @bganapa I gave Narayan internal links in Teams Chat, so didn't make sense to link them in a public discussion.

But the reason I am making this change, is because AAD graph is deprecated.

File: src/connectedk8s/azext_connectedk8s/custom.py
Comment (TheOnlyWei) [reply src/connectedk8s/azext_connectedk8s/custom.py:2259]: I am using the Azure CLI MS Graph client that they use for their RBAC module.

File: src/connectedk8s/azext_connectedk8s/custom.py
Comment (julienstroheker) [inline src/connectedk8s/azext_connectedk8s/custom.py:2261]: Should we handle the case where caller does not have `Application.Read.All` permission on graph ?

IMO this should be wrap into a `try` `catch` block and returns an actionable error message.

Comment (julienstroheker) [reply src/connectedk8s/azext_connectedk8s/custom.py:2261]: just checked the warning message the next line, we might just need to update the doc and add a prerequisites about this.

Remind me the current flow, if we do return `""` is the whole operation is failing ?

Comment (TheOnlyWei) [reply src/connectedk8s/azext_connectedk8s/custom.py:2261]: Custom Locations is enabled by default. If user doesn't have permission for retrieving the service principal via MS Graph, then it won't be enabled, and a warning message will be printed, which is the current behavior.

Comment (julienstroheker) [reply src/connectedk8s/azext_connectedk8s/custom.py:2261]: Make senses. Is just that would be great to provide more insights about the real failure here, if this is a graph permission issue or just that the call did returns no `oid`.
Can we ask PM to edit the doc : https://aka.ms/CustomLocationsObjectID to makes sure we ask customer to make sure they are having the `Application.Read.All` permission.

Comment (julienstroheker) [reply src/connectedk8s/azext_connectedk8s/custom.py:2261]: Actually I just checked and looks like `Application.Read.All` is granted by default on any default AAD User

Comment (TheOnlyWei) [reply src/connectedk8s/azext_connectedk8s/custom.py:2261]: For service principals, they need to be granted the permission explicitly by an admin.

**Review by bganapa — APPROVED** [review]

**Comment by yanzhudd** [comment]
please rerun related tests in live mode to verify this change.

**Comment by TheOnlyWei** [comment]
Hi @yanzhudd, we have already run the live tests.

**Review by yanzhudd — APPROVED** [review]

**Pull request merged**
